### PR TITLE
MGMT-18286: Internal Documentation for using and testing multipath

### PR DESCRIPTION
**Pull request by linoyaslan**
Create a detailed document outlining two approaches for configuring an environment with multipath.

## List all the issues related to this PR

- [ ] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [X] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [X] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: docs/dev/multipath.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+### STEP-2 - Create SAN VM + configure iSCSI target
+
+- Download the Fedora ISO to your Beaker machine.
+- In your VMM, create a new VM and select the ISO you downloaded, naming it "SAN".
+- Enable the root user and ssh
+- After installation is complete, go to "Show virtual hardware details" -> "Add Hardware" -> "Storage" -> create a 120GB disk.
+- SSH into your new VM and use `lsblk` to verify that you have a new block device in the list (likely named vdb).
+- Follow these instructions for setting up the iSCSI target (https://fedoraproject.org/wiki/Scsi-target-utils_Quickstart_Guide):
+    - Installing the scsi-target-utils package: `dnf -y install scsi-target-utils`
+    -  Start the firewalld service: `sudo systemctl start firewalld`
```
Comment (eliorerz) [inline docs/dev/multipath.md:22]:
```suggestion
    - Start the firewalld service: `sudo systemctl start firewalld`
```

File: docs/dev/multipath.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+### STEP-4 - 2nd Network as 2nd path to reach the iSCSI target
+
+1. In SAN VM - "show virtual hardware details" -> Add Hardware -> Network -> select the isolated network you created in STEP-1 -> copy the MAC address
+2. In SNO-SAN VM - "show virtual hardware details" -> Add Hardware -> Network -> select the isolated network you created in STEP-1 -> copy the MAC address
+3. In your beaker machine - 
+   - `sudo virsh net-edit SAN-network`
+   - In the dhcp section add the following:
+   
+   `<host mac='<SAN mac from point (1)>' name='SAN' ip='192.168.100.125'/>`
+    `<host mac='<SNO-SAN mac from point (2)>' name='SNO-SAN' ip='192.168.100.126'/>`
```
Comment (eliorerz) [inline docs/dev/multipath.md:49]:
```suggestion
   ```xml
   <host mac='<SAN mac from point (1)>' name='SAN' ip='192.168.100.125'/>
   <host mac='<SNO-SAN mac from point (2)>' name='SNO-SAN' ip='192.168.100.126'/>
   \```
```

File: docs/dev/multipath.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    - `sudo virsh net-destroy SAN-network`
+    - `sudo virsh net-start SAN-network`
+    - Check that the changes updated: `sudo virsh net-dumpxml SAN-network`
+4. Click "finish" in points 1-2.
+5. In SNO-SAN VM -
+    - `iscsiadm --mode node --targetname <iSCSI target you created in SAN> --portal 192.168.100.125:3260 -o new`
+    - `iscsiadm --mode node --targetname<iSCSI target you created in SAN> --portal 192.168.100.125:3260 --login`
```
Comment (eliorerz) [inline docs/dev/multipath.md:57]:
```suggestion
    - `iscsiadm --mode node --targetname <iSCSI target you created in SAN> --portal 192.168.100.125:3260 --login`
```

File: docs/dev/multipath.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+6. Run `multipath -ll` to confirm that multipathing is active.
+
+## 2nd approach
+
+In this approach, your beaker machine will act as the target, while the initiator will be the test-infra VM.
+
+1. SSH into your beaker machine and navigate to the directory where the test-infra is cloned.
+2. Create a VM that includes two different networks: `make deploy_nodes_with_networking MASTERS_COUNT=1`
+3. On the target side (beaker machine):
+   - create disk on hypervisor: `qemu-img create -o  preallocation=full -f qcow2 /tmp/disk0 30G`
```
Comment (eliorerz) [inline docs/dev/multipath.md:67]:
```suggestion
   - create disk on hypervisor: `qemu-img create -o preallocation=full -f qcow2 /tmp/disk0 30G`
```

File: docs/dev/multipath.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+   - Install tragetCli package: `dnf install targetcli -y`
+   - `targetcli clearconfig  confirm=True`
+   - `targetcli backstores/fileio create name=disk0 size=30G file_or_dev=/tmp/disk0`
+   - Create a target: `targetcli /iscsi create iqn.2023-01.com.example:target01`
+   - Create a LUN: `targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/luns create /backstores/fileio/disk0`
+   - Create portal:
+     - `targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <1st tearget IP>`
+     - `targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <2nd tearget IP>`
```
Comment (eliorerz) [inline docs/dev/multipath.md:75]:
```suggestion
     ```bash
     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <1st tearget IP>
     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <2nd tearget IP>
     ```
```

File: docs/dev/multipath.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+   - Allow access to initiator:
+     - `targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/acls create iqn.1994-05.com.redhat:e1c30923a3a`
+     - `targetcli / saveconfig`
```
Comment (eliorerz) [inline docs/dev/multipath.md:78]:
```suggestion
    ```bash
    targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/acls create iqn.1994-05.com.redhat:e1c30923a3a
    targetcli / saveconfig
    ```
```

File: docs/dev/multipath.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+4. On the node side (test-infra VM):
+   - SSH into the VM you created in step 2.
+   - Discover (client):
+     - `iscsiadm --mode discovery --type sendtargets --portal <1st target IP>`
+     - `iscsiadm --mode discovery --type sendtargets --portal <2nd target IP>`
```
Comment (eliorerz) [inline docs/dev/multipath.md:83]:
```suggestion
    ```bash
     iscsiadm --mode discovery --type sendtargets --portal <1st target IP>
     iscsiadm --mode discovery --type sendtargets --portal <2nd target IP>
```
```

File: docs/dev/multipath.md [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+   - Login to iSCSI:
+     - `iscsiadm -m node -T iqn.2023-01.com.example:target01 -p <1st target IP>:3260 -l`
+     - `iscsiadm -m node -T iqn.2023-01.com.example:target01 -p <2nd target IP>:3260 -l`
```
Comment (eliorerz) [inline docs/dev/multipath.md:86]:
```suggestion
    ```bash
     iscsiadm -m node -T iqn.2023-01.com.example:target01 -p <1st target IP>:3260 -l
     iscsiadm -m node -T iqn.2023-01.com.example:target01 -p <2nd target IP>:3260 -l
    ```
```

File: docs/dev/multipath.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+In this approach, your beaker machine will act as the target, while the initiator will be the test-infra VM.
+
+1. SSH into your beaker machine and navigate to the directory where the test-infra is cloned.
+2. Create a VM that includes two different networks: `make deploy_nodes_with_networking MASTERS_COUNT=1`
+3. On the target side (beaker machine):
+   - create disk on hypervisor: `qemu-img create -o  preallocation=full -f qcow2 /tmp/disk0 30G`
+   - Install tragetCli package: `dnf install targetcli -y`
+   - `targetcli clearconfig  confirm=True`
+   - `targetcli backstores/fileio create name=disk0 size=30G file_or_dev=/tmp/disk0`
```
Comment (eliorerz) [inline docs/dev/multipath.md:70]:
```suggestion
    ```bash
   targetcli clearconfig  confirm=True
   targetcli backstores/fileio create name=disk0 size=30G file_or_dev=/tmp/disk0
   ```
```

File: docs/dev/multipath.md (hunk 3 above)
Comment (eliorerz) [inline docs/dev/multipath.md:57]:
```suggestion
    ```bash
    iscsiadm --mode node --targetname <iSCSI target you created in SAN> --portal 192.168.100.125:3260 -o new
    iscsiadm --mode node --targetname<iSCSI target you created in SAN> --portal 192.168.100.125:3260 --login
    ```
```

**Review by eliorerz — CHANGES_REQUESTED** [review]
Nice work!
Look at my comments, I think that it would be better if the code will be in blocks and not in bullets

File: docs/dev/multipath.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+     ```bash
+     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <1st tearget IP>
+     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <2nd tearget IP>
+     ```
+   - Allow access to initiator:
+     ```bash
+     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/acls create iqn.1994-05.com.redhat:e1c30923a3a
+     targetcli / saveconfig
+     ```
+4. On the node side (test-infra VM):
```
Comment (AlonaKaplan) [inline docs/dev/multipath.md:203]: On the initiator side

File: docs/dev/multipath.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+   - SSH into the VM you created in step 2.
+   - Discover (client):
+     ```bash
+     iscsiadm --mode discovery --type sendtargets --portal <1st target IP>
+     iscsiadm --mode discovery --type sendtargets --portal <2nd target IP>
+     ```
+   - Login to iSCSI:
+     ```bash
+     iscsiadm -m node -T iqn.2023-01.com.example:target01 -p <1st target IP>:3260 -l
```
Comment (AlonaKaplan) [inline docs/dev/multipath.md:212]: In the first option there was no discovery (why?) and the syntax was slightly different (e.g --portal and not -p). Please keep it consistent.
 
```
 iscsiadm --mode node --targetname <iSCSI target you created in SAN> --portal 192.168.100.125:3260 -o new
    iscsiadm --mode node --targetname<iSCSI target you created in SAN> --portal 192.168.100.125:3260 --login
```

File: docs/dev/multipath.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+     ```bash
+     targetcli /iscsi create iqn.2023-01.com.example:target01
+     ```
+   - Create a LUN:
+     ```bash
+     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/luns create /backstores/fileio/disk0
+     ```
+   - Create portal:
+     ```bash
+     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <1st tearget IP>
```
Comment (AlonaKaplan) [inline docs/dev/multipath.md:224]: s/1st tearget IP/1st target IP
Where do you take the IP from? Is it the IP of the bridge the VM is connected to?

File: docs/dev/multipath.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+     targetcli /iscsi/iqn.2023-01.com.example:target01/tpg1/portals create <2nd tearget IP>
```
Comment (AlonaKaplan) [inline docs/dev/multipath.md:225]: same

File: docs/dev/multipath.md (hunk 12 above)
Comment (linoyaslan) [reply docs/dev/multipath.md:224]: @AlonaKaplan You can view the networks added to your beaker machine (which is the target) after running "make deploy_nodes_with_networking MASTERS_COUNT=1." To find the appropriate IP for each network, run "ip a" and search for the relevant network .

Comment (AlonaKaplan) [reply docs/dev/multipath.md]: I guess `ip a` on a beaker server provides a long output.  Can you please add more detailed instructions on how you "search for the relevant network".

Comment (linoyaslan) [reply docs/dev/multipath.md]: @AlonaKaplan  for example:
Screenshot from 2024-07-02 12-46-34

Comment (AlonaKaplan) [reply docs/dev/multipath.md]: I think it is the tap device, it's pretty weird it has an IP. I think the bridge IP should be used and not the tap one... But if it works maybe it doesn't matter, idk.

Comment (linoyaslan) [reply docs/dev/multipath.md]: @AlonaKaplan isn't the bridge IP is the one indicated by inet?

Comment (AlonaKaplan) [reply docs/dev/multipath.md]: I don't think it is the bridge, the bridge name usually has "virtbr" prefix.

**Comment by AlonaKaplan** [comment]
/lgtm

Nice! Thanks.

**Review by eliorerz — APPROVED** [review]
LGTM

**Pull request merged**